src/Common/src/TypeSystem/Common/TypeDesc.cs

```diff
@@ -103,9 +103,27 @@ public abstract partial class TypeDesc
 
         public override bool Equals(Object o)
         {
+            // Its only valid to compare two TypeDescs in the same context
+            Debug.Assert(o == null || !(o is TypeDesc) || Object.ReferenceEquals(((TypeDesc)o).Context, this.Context));
             return Object.ReferenceEquals(this, o);
         }
 
+#if DEBUG
+        public static bool operator ==(TypeDesc left, TypeDesc right)
```

Comment:
Well, this is just a set of asserts that should make this problematic case pretty easy to detect... I'll happily add the operators if you'd prefer.